waspc/src/Analyzer/TypeChecker/Internal.hs

```diff
@@ -0,0 +1,232 @@
+    annotateError k = left (\e -> UnificationError (ReasonDictWrongKeyType k e) typS typT)
+unifyTypes s t = Left $ UnificationError ReasonUncoercable s t
+
+-- | @weaken expr typ == Right expr'@ establishes that @expr@ can be typed as @typ@.
+-- Additionally, the following property is guaranteed:
+--
+-- * @exprType expr == typ@
+--
+-- When a @Left@ value is returned, then @expr@ can not be typed as @typ@.
+weaken :: Type -> TypedExpr -> Either TypeError TypedExpr
+weaken typ' expr
+  | exprType expr == typ' = Right expr
+-- A list can be weakened to @typ@ if
+-- - @typ@ is of the form @ListType typ'@
+-- - Every value in the list can be weakened to @typ'@
+weaken (ListType typ') expr@(List vals _) =
+  fmap (\values -> List values (ListType typ')) $
+    left (\e -> WeakenError (ReasonList e) expr typ') $
+      mapM (weaken typ') vals
+weaken (DictType typ') expr@(Dict entries _) = do
+  entries' <- mapM weakenEntry entries
+  mapM_ guardHasEntry $ M.toList typ'
+  return $ Dict entries' $ DictType typ'
+  where
+    -- Tries to apply [DictSome] and [DictNone] rules to the entries of the dict
+    weakenEntry :: (String, TypedExpr) -> Either TypeError (Ident, TypedExpr)
+    weakenEntry (key, value) = case M.lookup key typ' of
+      -- @key@ is missing from @typ'@ => extra keys are not allowed
+      Nothing -> Left $ WeakenError (ReasonDictExtraKey key) expr (DictType typ')
+      -- @key@ is required and present => only need to weaken the value's type
+      Just (DictRequired valueTyp) -> (key,) <$> annotateError key (weaken valueTyp value)
+      -- @key@ is optional and present => weaken value's type + use [DictSome]
+      Just (DictOptional valueTyp) -> (key,) <$> annotateError key (weaken valueTyp value)
+
+    -- Checks that all DictRequired entries in typ' exist in entries
+    guardHasEntry :: (String, DictEntryType) -> Either TypeError ()
+    guardHasEntry (key, DictOptional typ) = case lookup key entries of
+      -- @key@ is optional and missing => use [DictNone]
+      Nothing -> Right ()
+      -- @key@ is optional and present => weaken the value's type + use [DictSome]
+      Just entryVal -> case weaken typ entryVal of
+        Left e -> Left $ WeakenError (ReasonDictWrongKeyType key e) expr (DictType typ')
+        Right _ -> return ()
+    guardHasEntry (k, DictRequired t) = case lookup k entries of
+      -- @key@ is required and missing => not allowed
+      Nothing -> Left $ WeakenError (ReasonDictNoKey k) expr (DictType typ')
+      -- @key@ is required and present => only need to weaken value's type
+      Just entryVal -> case weaken t entryVal of
+        Left e -> Left $ WeakenError (ReasonDictWrongKeyType k e) expr (DictType typ')
+        Right _ -> return ()
+
+    -- Wraps a ReasonDictWrongKeyType error around a type error
+    annotateError :: String -> Either TypeError a -> Either TypeError a
+    annotateError k = left (\e -> WeakenError (ReasonDictWrongKeyType k e) expr (DictType typ'))
```

Comment:
There is some duplication between this line and lines above (2 places -> 214 and 221). Maybe you could reduce that duplication in some nice way? Probably use smth like `annotateError $ const () <$> weaken t entryVal` instead of using `case`?